File: live_trading/signals/decision_engine.py
"""
Decision Engine — Central orchestrator of all 7 signal engines.
Ported from decisionEngine.ts
"""
from dataclasses import dataclass, field
from typing import List, Literal, Optional
from live_trading.signals.gold_engine import OHLCV
from live_trading.signals.smc_engine import (
    SmcResult,
    analyze_smc_structure,
    detect_order_block_fake_breakout,
    get_latest_structure_event,
)
from live_trading.signals.wyckoff_engine import WyckoffResult, analyze_wyckoff
from live_trading.signals.price_action_engine import PriceActionResult, analyze_price_action
from live_trading.signals.trend_engine import TrendResult, analyze_trend
from live_trading.signals.market_regime import RegimeResult, RegimeEntryRules, detect_market_regime
from live_trading.signals.confidence_engine import ConfidenceResult, ConfidenceComponents, calc_confidence
from live_trading.signals.quality_filter import QualityFilterResult, apply_quality_filter, get_session_quality
from live_trading.signals.entry_filter import apply_entry_filter, EntryFilterResult
from live_trading.signals.range_strategy import RangeContext, evaluate_range_entry
from live_trading.signals.divergence_engine import analyze_divergence, DivergenceResult
from live_trading.risk.capital_manager import (
    FIXED_TP_RR, CapitalInput, CapitalOutput, calc_trade_parameters,
)
from live_trading.config import (
    CONF_HARD_MIN,
    RANGE_MIN_CONFIRMATIONS,
    REQUIRE_SMC_PRICE_ACTION_WYCKOFF,
)

# Marginal confidence R:R floor: trades with confidence between CONF_HARD_MIN
# and the regime minimum must still achieve this R:R to be allowed.
# 1.3 = profitable in expectancy even at 45% win rate (1.3 × 0.45 > 0.55).
CONF_MARGINAL_RR = 1.3

# Regimes where PA/Wyckoff naturally fire less often. Their existing adaptive
# one-step tightening is preserved; RANGE gets the explicit option-4 floor.
_CHOPPY_REGIMES = {"ACCUMULATION", "DISTRIBUTION", "HIGH_VOLATILITY"}


def _effective_min_confirmations(
    base_min_confirmations: int,
    regime: str,
    counter_trend: bool,
    range_min_confirmations: int = RANGE_MIN_CONFIRMATIONS,
) -> int:
    """Return the final N-of-4 gate without weakening the RANGE floor."""
    if regime == "RANGE":
        # Option 4 is a floor, not a replacement: a stricter operator setting
        # remains stricter. Counter-trend remains an additional requirement.
        range_floor = max(base_min_confirmations, range_min_confirmations)
        if counter_trend:
            return max(range_floor, min(base_min_confirmations + 1, 4))
        return range_floor
    if counter_trend:
        return min(base_min_confirmations + 1, 4)
    if regime in _CHOPPY_REGIMES:
        return min(base_min_confirmations + 1, 4)
    return base_min_confirmations


def _range_confirmation_gate(
    entry_filter: EntryFilterResult,
    min_confirmations: int,
) -> tuple[bool, str]:
    """Apply the RANGE-specific two-vote confirmation rule.

    RANGE entries always have an SMC direction because the candidate direction
    comes from SMC.  The second confirmation must be either Price Action or
    Wyckoff; a matching EMA trend alone is not sufficient in a choppy market.
    This intentionally overrides the global three-engine option-1 gate for
    RANGE only.  The separate edge, fresh sweep, reversal, R:R, and session
    limits remain mandatory.
    """
    if entry_filter.confirmation_count < min_confirmations:
        return (
            False,
            f"RANGE entry blocked: {entry_filter.confirmation_count}/"
            f"{min_confirmations} confirmations",
        )
    if not (entry_filter.price_action or entry_filter.wyckoff):
        return (
            False,
            "RANGE entry blocked: second confirmation must be "
            "Price Action or Wyckoff",
        )
    return True, ""


@dataclass
class DecisionResult:
    allowed:         bool
    direction:       Literal["BUY", "SELL", "NEUTRAL"]
    confidence:      float
    components:      ConfidenceComponents
    grade:           str
    regime:          str
    regime_label:    str
    regime_rules:    RegimeEntryRules
    quality_filter:  QualityFilterResult
    blocked_reasons: List[str]
    reasoning:       List[str]
    trade_params:    Optional[CapitalOutput]
    smc:    SmcResult
    wyckoff: WyckoffResult
    pa:     PriceActionResult
    trend:  TrendResult
    # Additive, optional — which of the 4 independent engines (SMC/Trend/
    # PriceAction/Wyckoff) voted for this trade's direction. None on the
    # early "no SMC signal" path, where the vote was never computed.
    # Existing callers that construct/consume DecisionResult are unaffected
    # since this has a default and nothing reads it unless it asks for it.
    entry_filter:    Optional[EntryFilterResult] = None
    divergence:      Optional[DivergenceResult]  = None
    dxy_signal:      str                         = "NEUTRAL"
    range_context:   Optional[RangeContext]      = None


def _candidate_direction(smc: SmcResult) -> str:
    # Use the newest event across both lists. Prioritising the last CHoCH
    # unconditionally can resurrect an old reversal against a newer BOS.
    latest_structure = get_latest_structure_event(smc)
    if latest_structure is not None:
        return latest_structure.type
    if smc.trend == "BULLISH": return "BUY"
    if smc.trend == "BEARISH": return "SELL"
    return "NEUTRAL"


def _make_neutral(
    smc, wyckoff, pa, trend, blocked_reasons, reasoning=None,
    range_context: Optional[RangeContext] = None,
) -> DecisionResult:
    from live_trading.signals.market_regime import REGIME_RULES
    rules = REGIME_RULES["RANGE"]
    return DecisionResult(
        allowed=False, direction="NEUTRAL", confidence=0.0,
        components=ConfidenceComponents(0,0,0,0,0,0,0),
        grade="REJECTED", regime="RANGE", regime_label="No Signal",
        regime_rules=rules,
        quality_filter=QualityFilterResult(
            allowed=False, blocked_reasons=blocked_reasons,
            session_quality="BLOCKED", adx=0.0,
            is_severe_range=False, is_late_entry=False,
            is_low_probability=False, is_fake_breakout=False,
            is_weak_volume=False, is_low_momentum=False,
        ),
        blocked_reasons=blocked_reasons,
        reasoning=reasoning or [],
        trade_params=None,
        smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
        range_context=range_context,
    )


def run_decision_engine(
    candles:           List[OHLCV],
    account_balance:   float,
    risk_percent:      float = 1.0,
    min_confirmations: int   = 1,
    use_atr_high_vol:  bool  = False,
    dxy_signal:        str   = "NEUTRAL",
    require_price_action: bool = False,
    require_smc_price_action_wyckoff: bool = REQUIRE_SMC_PRICE_ACTION_WYCKOFF,
    range_trading_enabled: bool = True,
    range_min_confirmations: int = 2,
    range_min_rr: float = 1.5,
    range_edge_atr_distance: float = 0.25,
    range_risk_percent: Optional[float] = None,
) -> DecisionResult:

    smc     = analyze_smc_structure(candles)
    wyckoff = analyze_wyckoff(candles)
    pa      = analyze_price_action(candles)
    trend   = analyze_trend(candles)

    candidate = _candidate_direction(smc)
    if candidate == "NEUTRAL":
        return _make_neutral(smc, wyckoff, pa, trend, ["No SMC signal"])

    # Soft EMA gate — counter-trend trades are allowed but need 3 confirmations
    trend_dir = ("BUY" if trend.trend == "BULLISH" else
                 "SELL" if trend.trend == "BEARISH" else "NEUTRAL")
    _counter_trend = (candidate == "BUY" and trend_dir == "SELL") or \
                     (candidate == "SELL" and trend_dir == "BUY")

    # Detect regime early — needed to set the adaptive confirmation threshold.
    regime = detect_market_regime(candles, trend, wyckoff, use_atr_high_vol)
    effective_min_confirmations = _effective_min_confirmations(
        min_confirmations,
        regime.regime,
        _counter_trend,
    )

    # Entry filter — minimum N-of-4 vote gate (SMC always required).
    # RANGE has a narrower rule than ordinary regimes: SMC plus either
    # Price Action or Wyckoff is sufficient; the global option-1 gate must
    # not turn that dedicated two-confirmation playbook into a three-vote gate.
    is_range_regime = regime.regime == "RANGE"
    ef = apply_entry_filter(
        smc_signal      = candidate,
        ema_trend       = trend.trend,
        pa_signal       = pa.pa_signal,
        wyckoff_signal  = wyckoff.wyckoff_signal,
        min_confirmations = effective_min_confirmations,
        require_price_action = require_price_action and not is_range_regime,
        require_smc_price_action_wyckoff = (
            require_smc_price_action_wyckoff and not is_range_regime
        ),
    )
    if is_range_regime:
        range_votes_ok, range_votes_reason = _range_confirmation_gate(
            ef, range_min_confirmations
        )
        if not range_votes_ok:
            return _make_neutral(
                smc, wyckoff, pa, trend,
                [range_votes_reason],
                [range_votes_reason],
            )
    elif not ef.allowed:
        votes = (f"SMC={'✓' if ef.smc else '✗'}  "
                 f"Trend={'✓' if ef.trend else '✗'}  "
                 f"PA={'✓' if ef.price_action else '✗'}  "
                 f"Wyckoff={'✓' if ef.wyckoff else '✗'}")
        if (
            require_smc_price_action_wyckoff
            and not (ef.smc and ef.price_action and ef.wyckoff)
        ):
            reason = (
                "Entry filter: Option 1 requires SMC + Price Action + Wyckoff — "
                f"{votes}  [regime={regime.regime}]"
            )
        elif require_price_action and not ef.price_action:
            reason = (f"Entry filter: Price Action confirmation required — "
                      f"{votes}  [regime={regime.regime}]")
        else:
            reason = (f"Entry filter: only {ef.confirmation_count}/{effective_min_confirmations} "
                      f"confirmations — {votes}  [regime={regime.regime}]")
        return _make_neutral(smc, wyckoff, pa, trend, [reason], [reason])

    # RANGE is a separate playbook. It must never inherit a trend entry just
    # because the generic four-engine vote happened to pass.
    range_context: Optional[RangeContext] = None
    if is_range_regime:
        if not range_trading_enabled:
            return _make_neutral(
                smc, wyckoff, pa, trend,
                ["RANGE trading is disabled"],
                ["RANGE trading is disabled"],
            )
        range_context = evaluate_range_entry(
            candles=candles,
            direction=candidate,
            smc=smc,
            pa=pa,
            confirmation_count=ef.confirmation_count,
            min_confirmations=range_min_confirmations,
            edge_atr_distance=range_edge_atr_distance,
        )
        if not range_context.valid:
            return _make_neutral(
                smc, wyckoff, pa, trend,
                [range_context.reason],
                [range_context.reason],
                range_context=range_context,
            )

    if candidate == "BUY"  and not regime.rules.allow_long:
        return _make_neutral(smc, wyckoff, pa, trend,
                             [f'Regime "{regime.rules.label}" does not allow LONG'])
    if candidate == "SELL" and not regime.rules.allow_short:
        return _make_neutral(smc, wyckoff, pa, trend,
                             [f'Regime "{regime.rules.label}" does not allow SHORT'])

    last_candle  = candles[-1]
    session      = get_session_quality(last_candle.time)
    divergence   = analyze_divergence(candles)
    # Option 3: DXY is retained as telemetry only and cannot affect entry
    # confidence or the decision. The confidence engine explicitly ignores
    # this legacy compatibility argument.
    conf_result  = calc_confidence(
        smc, wyckoff, pa, trend, regime, session, candidate,
        divergence_signal=divergence.signal,
    )

    if conf_result.confidence < CONF_HARD_MIN:
        n = DecisionResult(
            allowed=False, direction=candidate,  # type: ignore
            confidence=conf_result.confidence, components=conf_result.components,
            grade="REJECTED", regime=regime.regime, regime_label=regime.rules.label,
            regime_rules=regime.rules,
            quality_filter=QualityFilterResult(
                False, [f"Confidence {conf_result.confidence:.1f}% < {CONF_HARD_MIN}% minimum"],
                session, regime.adx, False, False, True, False, False, False),
            blocked_reasons=[f"Confidence {conf_result.confidence:.1f}% < {CONF_HARD_MIN}%"],
            reasoning=conf_result.reasoning, trade_params=None,
            smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
            entry_filter=ef,
        )
        return n

    latest_structure = get_latest_structure_event(smc)
    last_structure_bar = (latest_structure.bar_index
                          if latest_structure is not None else None)
    # Feed the newest BOS/CHoCH bar through the existing quality-filter slot
    # so both event types share the same freshness gate.
    quality  = apply_quality_filter(candles, candidate, conf_result.confidence,
                                    last_structure_bar, regime.adx, regime.atr_ratio)
    if not quality.allowed:
        return DecisionResult(
            allowed=False, direction=candidate,  # type: ignore
            confidence=conf_result.confidence, components=conf_result.components,
            grade=conf_result.grade, regime=regime.regime, regime_label=regime.rules.label,
            regime_rules=regime.rules, quality_filter=quality,
            blocked_reasons=quality.blocked_reasons, reasoning=conf_result.reasoning,
            trade_params=None, smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
            entry_filter=ef,
        )

    # Option 2 hard gate: a directional breakout that closes back inside its
    # aligned Order Block is treated as a fake breakout.  Do not let this
    # setup reach capital sizing or the order executor.  The existing quality
    # result is reused so the panel receives the normal filter telemetry plus
    # the explicit rejection flag.
    fake_ob = detect_order_block_fake_breakout(candles, smc, candidate)
    if fake_ob is not None:
        fake_reason = (
            f"Fake Breakout in {fake_ob.type.title()} Order Block "
            f"[{fake_ob.low:.2f}, {fake_ob.high:.2f}] — entry blocked"
        )
        quality.allowed = False
        quality.is_fake_breakout = True
        quality.blocked_reasons.append(fake_reason)
        return DecisionResult(
            allowed=False, direction=candidate,  # type: ignore
            confidence=conf_result.confidence, components=conf_result.components,
            grade=conf_result.grade, regime=regime.regime,
            regime_label=regime.rules.label, regime_rules=regime.rules,
            quality_filter=quality,
            blocked_reasons=[fake_reason],
            reasoning=conf_result.reasoning + [fake_reason],
            trade_params=None,
            smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
            entry_filter=ef,
            divergence=divergence,
            dxy_signal=dxy_signal,
        )

    # Capital manager inputs
    aligned_obs = [ob for ob in smc.order_blocks
                   if ob.type == ("BULLISH" if candidate == "BUY" else "BEARISH")]
    latest_ob = aligned_obs[-1] if aligned_obs else None

    entry = last_candle.close

    # H-1 FIX: use most-recent directionally-valid BOS price as the SL anchor,
    # not the global max/min across all time.
    # BUY SL anchor: most recent SELL-BOS price below entry (= broken swing low)
    # SELL SL anchor: most recent BUY-BOS price above entry (= broken swing high)
    sell_bos_below = [b.price for b in smc.bos_signals if b.type == "SELL" and b.price < entry]
    buy_bos_above  = [b.price for b in smc.bos_signals if b.type == "BUY"  and b.price > entry]

    # H-2 FIX: populate support/resistance from SMC equal levels (previously always None).
    # Equal lows = institutional demand / support; equal highs = supply / resistance.
    eq_support    = (smc.equal_lows[-1].price
                     if smc.equal_lows  and smc.equal_lows[-1].price  < entry else None)
    eq_resistance = (smc.equal_highs[-1].price
                     if smc.equal_highs and smc.equal_highs[-1].price > entry else None)
    range_support = (
        range_context.support
        if range_context is not None and range_context.support < entry
        else None
    )
    range_resistance = (
        range_context.resistance
        if range_context is not None and range_context.resistance > entry
        else None
    )

    cap_input = CapitalInput(
        direction=candidate,
        entry_price=entry,
        atr=regime.atr,
        account_balance=account_balance,
        risk_percent=(
            range_risk_percent
            if regime.regime == "RANGE" and range_risk_percent is not None
            else risk_percent
        ),
        take_profit_rr=range_min_rr if regime.regime == "RANGE" else FIXED_TP_RR,
        order_block_top=latest_ob.high if latest_ob else None,
        order_block_bottom=latest_ob.low if latest_ob else None,
        swing_high=buy_bos_above[-1]  if buy_bos_above  else None,
        swing_low=sell_bos_below[-1]  if sell_bos_below else None,
        support_level=range_support or eq_support,
        resistance_level=range_resistance or eq_resistance,
        take_profit_level=(
            range_resistance if candidate == "BUY" else range_support
        ),
    )
    trade_params = calc_trade_parameters(cap_input)

    # Marginal confidence check
    min_conf = regime.rules.min_confidence
    if conf_result.confidence < min_conf:
        if trade_params.risk_reward_ratio < CONF_MARGINAL_RR:
            return DecisionResult(
                allowed=False, direction=candidate,  # type: ignore
                confidence=conf_result.confidence, components=conf_result.components,
                grade="MARGINAL", regime=regime.regime, regime_label=regime.rules.label,
                regime_rules=regime.rules, quality_filter=quality,
                blocked_reasons=[
                    f"Marginal conf {conf_result.confidence:.1f}% requires R:R ≥ {CONF_MARGINAL_RR} "
                    f"(got {trade_params.risk_reward_ratio:.2f})"
                ],
                reasoning=conf_result.reasoning, trade_params=None,
                smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
                entry_filter=ef,
            )

    # R:R gate
    required_rr = range_min_rr if regime.regime == "RANGE" else regime.rules.min_rr
    if trade_params.risk_reward_ratio < required_rr:
        return DecisionResult(
            allowed=False, direction=candidate,  # type: ignore
            confidence=conf_result.confidence, components=conf_result.components,
            grade=conf_result.grade, regime=regime.regime, regime_label=regime.rules.label,
            regime_rules=regime.rules, quality_filter=quality,
            blocked_reasons=[
                f"R:R {trade_params.risk_reward_ratio:.2f} < {required_rr} "
                f"minimum for {regime.rules.label}"
            ],
            reasoning=conf_result.reasoning, trade_params=None,
            smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
            entry_filter=ef,
        )

    # ── TRADE ALLOWED ─────────────────────────────────────────────────────────
    return DecisionResult(
        allowed=True, direction=candidate,  # type: ignore
        confidence=conf_result.confidence, components=conf_result.components,
        grade=conf_result.grade, regime=regime.regime, regime_label=regime.rules.label,
        regime_rules=regime.rules, quality_filter=quality,
        blocked_reasons=[], reasoning=conf_result.reasoning,
        trade_params=trade_params,
        smc=smc, wyckoff=wyckoff, pa=pa, trend=trend,
        entry_filter=ef,
        divergence=divergence,
        dxy_signal=dxy_signal,
        range_context=range_context,
    )


def describe_strategy(decision: "DecisionResult") -> dict:
    """Build a human-readable summary of *why* this trade was taken.

    Purely derived from data the decision engine already computed — it adds
    no new signal logic and cannot change whether a trade is taken. Intended
    to travel alongside a just-opened trade (e.g. published to Redis by the
    live loop) so the Telegram panel can explain the trade in its
    "TRADE OPENED" notification instead of showing only price/volume/SL/TP.
    """
    ef = decision.entry_filter
    _ENGINE_NAMES = {
        "smc":          "Smart Money Concepts (structure)",
        "trend":        "Trend (EMA alignment)",
        "price_action": "Price Action",
        "wyckoff":      "Wyckoff",
    }
    if ef is not None:
        confirmations = [
            label for key, label in _ENGINE_NAMES.items() if getattr(ef, key)
        ]
        confirmation_count = ef.confirmation_count
    else:
        confirmations = []
        confirmation_count = 0

    return {
        "direction":           decision.direction,
        "grade":               decision.grade,
        "confidence":          round(decision.confidence, 1),
        "regime":              decision.regime,
        "regime_label":        decision.regime_label,
        "confirmations":       confirmations,
        "confirmation_count":  confirmation_count,
        "confirmation_total":  4,
        # Top signal-level reasons behind the confidence score (e.g. "BOS
        # confirmed", "Strong EMA alignment (50/100/200)", "Spring confirmed").
        "signals":             list(decision.reasoning[:6]),
    }

File: robot/src/lib/decisionEngine.ts
// Decision Engine — Central Orchestrator (v4.0)
//
// Single entry point that connects ALL 7 engines:
//   ① SMC Engine         — structure, BOS/CHoCH, OBs, FVGs, sweeps
//   ② Wyckoff Engine     — phase, Spring/Upthrust, volume
//   ③ Price Action Engine — patterns, levels, breakouts
//   ④ Trend Engine       — EMA 50/100/200 gate
//   ⑤ Market Regime Detector — adaptive entry rules per regime
//   ⑥ Confidence Engine  — 0–100% weighted score
//   ⑦ Quality Filter     — 10-category final gate
//   ⑧ Capital Manager    — SL/TP/lots from structural levels
//
// No trade decision is made anywhere else in the codebase.
// All engines contribute scores; no single engine can block or open
// a trade on its own.

import { OHLCV } from './goldEngine.js';
import { analyzeSmcStructure, type SmcResult } from './smcEngine.js';
import { analyzeWyckoff, type WyckoffResult } from './wyckoffEngine.js';
import { analyzePriceAction, type PriceActionResult } from './priceActionEngine.js';
import { analyzeTrend, type TrendResult } from './trendEngine.js';
import { detectMarketRegime, type RegimeResult, type MarketRegime } from './marketRegimeDetector.js';
import { calcConfidence, type ConfidenceResult, type ConfidenceComponents } from './confidenceEngine.js';
import { applyQualityFilter, getSessionQuality, type QualityFilterResult } from './qualityFilter.js';
import { calcTradeParameters, type CapitalInput, type CapitalOutput } from './capitalManager.js';

// ===== TYPES =====

export interface DecisionResult {
  /** Final gate — only true when ALL checks pass and confidence meets threshold */
  allowed:           boolean;
  direction:         'BUY' | 'SELL' | 'NEUTRAL';
  /** Confidence percentage, e.g. 91.8. 0 when no signal. */
  confidence:        number;
  components:        ConfidenceComponents;
  grade:             ConfidenceResult['grade'];
  regime:            MarketRegime;
  regimeLabel:       string;
  regimeRules:       RegimeResult['rules'];

  // Quality gate details
  qualityFilter:     QualityFilterResult;
  blockedReasons:    string[];

  // Decision reasoning (bullet points for logging/UI)
  reasoning:         string[];

  // Trade parameters (only populated when allowed === true)
  tradeParams?:      CapitalOutput;

  // Raw engine outputs (for API transparency)
  smc:    SmcResult;
  wyckoff: WyckoffResult;
  pa:     PriceActionResult;
  trend:  TrendResult;
}

// ===== CONSTANTS =====

/**
 * Hard minimum confidence — trades below this are ALWAYS rejected.
 *
 * Empirical calibration note (XAUUSD M5, 2021–2025):
 *   The original 85% threshold assumed synthetic / tick-level data where the
 *   Wyckoff, SMC, and PA engines all score near their theoretical maxima
 *   simultaneously.  On real M5 aggregated data the empirical maximum across
 *   340 k bars is 84.4%, which proves 85% is mathematically unattainable.
 *
 *   After auto-calibrating CFG_M5 (maxRangePct from data percentile, spring-
 *   margin from median ATR), fixing the spring-detection bug, and reducing the
 *   trendPct gate, the engine was re-validated with a full 5-year CSV backtest.
 *   The threshold is lowered to 70% to capture the populated 70–84% confidence
 *   band while still demanding strong multi-component alignment.
 */
const CONF_HARD_MIN  = 70;
/** Marginal zone: 70–regime.minConf — allowed when R:R ≥ CONF_MARGINAL_RR.
 *  Lowered from 2.0 to 1.5 to align with STRONG_TREND regime's minRR=1.5. */
const CONF_MARGINAL_RR = 1.5;

// ===== SESSION HELPER =====
// getSessionQuality is re-exported from qualityFilter.ts and imported above.
// The previous local copy was an exact duplicate of that export; removed to
// eliminate the maintenance risk of the two implementations diverging silently.

// ===== CANDIDATE DIRECTION =====
// Priority: CHoCH > BOS > structural trend.
// This is the direction the confidence engine evaluates.

function candidateDirection(smc: SmcResult): 'BUY' | 'SELL' | 'NEUTRAL' {
  const lastChoch = smc.chochSignals[smc.chochSignals.length - 1];
  if (lastChoch) return lastChoch.type;

  const lastBos = smc.bosSignals[smc.bosSignals.length - 1];
  if (lastBos) return lastBos.type;

  if (smc.trend === 'BULLISH') return 'BUY';
  if (smc.trend === 'BEARISH') return 'SELL';

  return 'NEUTRAL';
}

// ===== MAIN ENTRY POINT =====

// ===== ENGINE CONFIG =====
// Optional configuration for comparative backtesting.
// minConfirmations: 2 or 3 (default 3)
// useAtrHighVolFilter: enable/disable ATR spike regime (default false — disabled)
export interface DecisionEngineConfig {
  minConfirmations?:    number;   // default: 3
  useAtrHighVolFilter?: boolean;  // default: false (disabled — reduced profits in backtests)
}

export function runDecisionEngine(
  candles:        OHLCV[],
  timeframe:      'M1' | 'M5',
  accountBalance: number,
  riskPercent = 1.0,
  engConfig: DecisionEngineConfig = {},
): DecisionResult {
  const minConfs       = engConfig.minConfirmations   ?? 3;
  const useAtrHighVol  = engConfig.useAtrHighVolFilter ?? false;

  // ── Run all 4 core engines ────────────────────────────────────────
  // analyzeSmcStructure internally calls Wyckoff, PA, Trend for its own
  // scoring.  We call them separately here so the Decision Engine has
  // access to the raw outputs for its own confidence calculation.
  const smc     = analyzeSmcStructure(candles, timeframe, minConfs);
  const wyckoff = analyzeWyckoff(candles, timeframe);
  const pa      = analyzePriceAction(candles, timeframe);
  const trend   = analyzeTrend(candles);

  // ── Determine candidate direction ─────────────────────────────────
  const candidate = candidateDirection(smc);

  const neutral: DecisionResult = {
    allowed:        false,
    direction:      'NEUTRAL',
    confidence:     0,
    components:     { smcScore: 0, trendScore: 0, paScore: 0, wyckoffScore: 0, liquidityScore: 0, volatilityScore: 0, total: 0 },
    grade:          'REJECTED',
    regime:         'RANGE',
    regimeLabel:    'No Signal',
    regimeRules:    { minConfidence: 90, minRR: 1.5, allowLong: true, allowShort: true, slAtrMultAdjust: 1, label: 'No Signal' },
    qualityFilter:  { allowed: false, blockedReasons: ['No SMC signal'], sessionQuality: 'BLOCKED', adx: 0, isSevereRange: false, isLateEntry: false, isLowProbability: false, isFakeBreakout: false, isWeakVolume: false, isLowMomentum: false },
    blockedReasons: ['No SMC signal'],
    reasoning:      [],
    smc, wyckoff, pa, trend,
  };

  if (candidate === 'NEUTRAL') return neutral;

  // ── Hard EMA gate (direct trend opposition blocks immediately) ────
  const trendDir = trend.trend === 'BULLISH' ? 'BUY' : trend.trend === 'BEARISH' ? 'SELL' : 'NEUTRAL';
  if ((candidate === 'BUY' && trendDir === 'SELL') || (candidate === 'SELL' && trendDir === 'BUY')) {
    return {
      ...neutral,
      blockedReasons: [`EMA trend (${trend.trend}) directly opposes SMC signal (${candidate})`],
      reasoning:      [`EMA trend (${trend.trend}) directly opposes SMC signal (${candidate})`],
    };
  }

  // ── Market Regime ─────────────────────────────────────────────────
  const regime = detectMarketRegime(candles, trend, wyckoff, useAtrHighVol);

  // ── Direction allowed by regime? ──────────────────────────────────
  if (candidate === 'BUY'  && !regime.rules.allowLong) {
    return {
      ...neutral,
      regime:      regime.regime,
      regimeLabel: regime.rules.label,
      regimeRules: regime.rules,
      blockedReasons: [`Regime "${regime.rules.label}" does not allow LONG entries`],
      reasoning:      [`Regime: ${regime.description}`],
    };
  }
  if (candidate === 'SELL' && !regime.rules.allowShort) {
    return {
      ...neutral,
      regime:      regime.regime,
      regimeLabel: regime.rules.label,
      regimeRules: regime.rules,
      blockedReasons: [`Regime "${regime.rules.label}" does not allow SHORT entries`],
      reasoning:      [`Regime: ${regime.description}`],
    };
  }

  // ── Session quality (feeds both confidence engine and quality filter) ─
  const lastCandle = candles[candles.length - 1];
  const sessionQuality = getSessionQuality(lastCandle.time);

  // ── Confidence Score Engine ───────────────────────────────────────
  const confResult = calcConfidence(smc, wyckoff, pa, trend, regime, sessionQuality, candidate);

  // ── Hard minimum confidence check ────────────────────────────────
  if (confResult.confidence < CONF_HARD_MIN) {
    return {
      ...neutral,
      direction:      candidate,
      confidence:     confResult.confidence,
      components:     confResult.components,
      grade:          'REJECTED',
      regime:         regime.regime,
      regimeLabel:    regime.rules.label,
      regimeRules:    regime.rules,
      blockedReasons: [`Confidence ${confResult.confidence.toFixed(1)}% < ${CONF_HARD_MIN}% hard minimum`],
      reasoning:      confResult.reasoning,
      smc, wyckoff, pa, trend,
    };
  }

  // ── Quality Filter (10-category gate) ────────────────────────────
  const lastBosBarIndex = smc.bosSignals.length > 0
    ? smc.bosSignals[smc.bosSignals.length - 1].barIndex
    : null;
  const quality = applyQualityFilter(candles, candidate, confResult.confidence, lastBosBarIndex, regime.adx, regime.atrRatio);

  if (!quality.allowed) {
    return {
      ...neutral,
      direction:      candidate,
      confidence:     confResult.confidence,
      components:     confResult.components,
      grade:          confResult.grade,
      regime:         regime.regime,
      regimeLabel:    regime.rules.label,
      regimeRules:    regime.rules,
      qualityFilter:  quality,
      blockedReasons: quality.blockedReasons,
      reasoning:      confResult.reasoning,
      smc, wyckoff, pa, trend,
    };
  }

  // ── Capital Manager ───────────────────────────────────────────────
  const lastCanc = candles[candles.length - 1];
  const atr = regime.atr;

  // Best structural levels for SL/TP
  const alignedObs = smc.orderBlocks.filter(ob =>
    ob.type === (candidate === 'BUY' ? 'BULLISH' : 'BEARISH'));
  const latestOb = alignedObs[alignedObs.length - 1];

  const swingHighs = smc.bosSignals.filter(b => b.type === 'BUY').map(b => b.price);
  const swingLows  = smc.bosSignals.filter(b => b.type === 'SELL').map(b => b.price);

  const capitalInput: CapitalInput = {
    direction:         candidate,
    entryPrice:        lastCanc.close,
    atr,
    accountBalance,
    riskPercent,
    orderBlockTop:     latestOb?.high,
    orderBlockBottom:  latestOb?.low,
    swingHigh:         swingHighs.length ? Math.max(...swingHighs) : undefined,
    swingLow:          swingLows.length  ? Math.min(...swingLows)  : undefined,
  };
  const tradeParams = calcTradeParameters(capitalInput);

  // ── Marginal confidence check (85–threshold needs R:R ≥ 2.0) ─────
  const minConf = regime.rules.minConfidence;
  if (confResult.confidence < minConf) {
    if (tradeParams.riskRewardRatio < CONF_MARGINAL_RR) {
      return {
        ...neutral,
        direction:      candidate,
        confidence:     confResult.confidence,
        components:     confResult.components,
        grade:          'MARGINAL',
        regime:         regime.regime,
        regimeLabel:    regime.rules.label,
        regimeRules:    regime.rules,
        qualityFilter:  quality,
        blockedReasons: [
          `Marginal confidence ${confResult.confidence.toFixed(1)}% (< ${minConf}%) requires R:R ≥ ${CONF_MARGINAL_RR} (got ${tradeParams.riskRewardRatio.toFixed(2)})`,
        ],
        reasoning: confResult.reasoning,
        smc, wyckoff, pa, trend,
      };
    }
  }

  // ── R:R gate (regime-specific minimum) ───────────────────────────
  if (tradeParams.riskRewardRatio < regime.rules.minRR) {
    return {
      ...neutral,
      direction:      candidate,
      confidence:     confResult.confidence,
      components:     confResult.components,
      grade:          confResult.grade,
      regime:         regime.regime,
      regimeLabel:    regime.rules.label,
      regimeRules:    regime.rules,
      qualityFilter:  quality,
      blockedReasons: [
        `R:R ${tradeParams.riskRewardRatio.toFixed(2)} < ${regime.rules.minRR} minimum for "${regime.rules.label}" regime`,
      ],
      reasoning:      confResult.reasoning,
      smc, wyckoff, pa, trend,
    };
  }

  // ── TRADE ALLOWED ─────────────────────────────────────────────────
  return {
    allowed:        true,
    direction:      candidate,
    confidence:     confResult.confidence,
    components:     confResult.components,
    grade:          confResult.grade,
    regime:         regime.regime,
    regimeLabel:    regime.rules.label,
    regimeRules:    regime.rules,
    qualityFilter:  quality,
    blockedReasons: [],
    reasoning:      confResult.reasoning,
    tradeParams,
    smc, wyckoff, pa, trend,
  };
}
